Enhance `Utils.kt` with common elements and add Android tests

NBCCF025 opened the pull request:
**Description:**  
This PR improves the `Utils.kt` file by adding several common utility functions that can be reused across the app. Additionally, it includes comprehensive Android tests for these new utilities within `UtilsTest.kt`, ensuring reliability and expected behavior on Android devices. The purpose of this functions is to reduce code duplication, to have a design basis and keep the code clean. Related task #162 , #163 .


**Changes Made:**
- Added new common utility functions to `Utils.kt` for better code reusability.
- Created Android tests in `UtilsTest.kt` to validate the functionality and robustness of the new utility functions.
- Updated documentation for all functions.


**Testing:**
- All new functions in `Utils.kt` have corresponding Android tests in `UtilsTest.kt`.
- Verified that tests pass across different device configurations to ensure compatibility and accuracy.


**Understanding The Changes Made:**
- ScaffoldMainScreen() is a function that should be called for the following screens:
	- ChallengeScreen.kt
	- Home.kt
	- Profile.kt
It create the basis of the screen, so that it already contain common elements such as a default column, the top blue bar, the Bottom navigation menu and the information button with the gestion of the popup. The specific elements for each screen should be passed as a lambda.

- ScaffoldAnnexeScreen() is a function that should be called for the Following screens:
	- ChallengeHistoryScreen
	- CreateAChallengeScreen
	- NewChallengeScreen
	- ExerciseScreenEasy
	- ExerciseScreenHard
	- QuestScreen
	- FriendsListScreen
	- MyStatsScreen
	- SettingsScreen
It create the basis of the screen, so that it already contain common elements such as a default column, the top blue bar and the back button. The specific elements for each screen should be passed as a lambda.

- ReusableButtonWithIcon(), ReusableTextButton(), SquareButton() and UtilButton() are still the same and can be used in the screens for their specific part.

- StatisticsRow(), AllLetterLearned(), StreakCounter(), ProfilePicture(), AccountInformation() and NotImplementedYet() are function that can be used in the screens for their specific part.

- TopBar(), BottomBar(), BackButton(), ColumnScreen(), InfoPopup() and HorizontalLetterList() are function that shouldn't be used directly in the screens as they are helper function or already called in ScaffoldMainScreen()/ScaffoldAnnexeScreen().


**Future task and improvement:**
This PR allow us to make new tasks with a clear vision in order to reduce code duplication and to have a common design for all screens.

ThomasMPont commented:
I'm confused about how I'm supposed to tell (or you for that matter) whether these new elements look and work the way they should. Have you looked at these by hand locally?

I suggest you use at least 1 instance of each of these elements somewhere on the app, so that we have something to review. This could be listed as a separate task, and even done by someone else if you're out of time.

NBCCF025 commented:
**Functions**

The following functions are already implemented and used in the screens :
- ReusableButtonWithIcon()
- ReusableTextButton()
- SquareButton()
- UtilButton()
- StreakCounter()
- ProfilePicture()
- TopBar()
- BackButton()
- InfoPopup()
- HorizontalLetterList()

The following functions are just combination of them that we can already see on the screens :
- BottomBar()
- AllLetterLearned()
- AccountInformation()

The following function only have a mockup in the test :
- StatisticsRow()
- NotImplementedYet()
- ColumnScreen()
- ScaffoldMainScreen()
- ScaffoldAnnexeScreen()

**Small PR**
As we want to have small PR, I haven't done any modification on the screens for now, I wanted to let that for latter tasks and if needed, improvement on the functions that are not currently used could be done at the same time.

ziyadbagh commented:
> **Functions**
> 
> The following functions are already implemented and used in the screens :
> 
> * ReusableButtonWithIcon()
> * ReusableTextButton()
> * SquareButton()
> * UtilButton()
> * StreakCounter()
> * ProfilePicture()
> * TopBar()
> * BackButton()
> * InfoPopup()
> * HorizontalLetterList()
> 
> The following functions are just combination of them that we can already see on the screens :
> 
> * BottomBar()
> * AllLetterLearned()
> * AccountInformation()
> 
> The following function only have a mockup in the test :
> 
> * StatisticsRow()
> * NotImplementedYet()
> * ColumnScreen()
> * ScaffoldMainScreen()
> * ScaffoldAnnexeScreen()
> 
> **Small PR** As we want to have small PR, I haven't done any modification on the screens for now, I wanted to let that for latter tasks and if needed, improvement on the functions that are not currently used could be done at the same time.

ziyadbagh commented:
I'm struggling to understand the concept of a small PR and the emphasis on keeping them minimal. To me, a PR should make a meaningful impact on the code and serve a clear purpose. Personally, I don't mind large PRs, as long as they're well-documented and thoroughly tested.

NBCCF025 commented:
Ok, so I suggest we close this PR and we do the refactor of two screens to ensure all the functions work as intended.

I have created two new tasks #167 and #168 that correspond respectively to MyStatsScreen (Annexe screen) and ProfileScreen (Main screen).

When it will be done I will make a new PR with all the changes.

ziyadbagh commented:
Please wait until we've heard from everyone else, so we're all on the same page.

ThomasMPont reviewed (changes requested):
From what I've gathered from our discussion on Whatsapp, we have decided not to merge this branch until all screens have been refactored with the new elements. These refactors should be done in new tasks on this branch.

This PR will be reevaluated once all screens have been refactored.

SaFouNaldo commented:
@NBCCF025 I guess you can mark the related issue as done in S4 so that the TA's can copute the overall implication in the sprint. 
As Thomas stayted we can wait for meging in if there is no inconvinience

NBCCF025 commented:
Related tasks: #162 , #163 , #167 , #168 .

**Changes Made:**
- Added new common utility functions to Utils.kt for better code reusability.
- Created Android tests in UtilsTest.kt to validate the functionality and robustness of the new utility functions.
- Updated documentation for all functions.
- Modified and tested the MyStatsScreen.kt to use the new functions and addapt to the new figma design
- Modified and tested the ProfileScreen.kt to use the new functions

**Utils Functions:**
- All the functions in Utils.kt are now used within the app and tested.

**Future Enhancement:**
- Other screens could be refactor to use common functions.

**SonarCloud Issue:**
- The issue raised by sonarcloud is about MyStatsScreen() that take 8 argument and this will be changed in the future when we will directly passed the user as the arguement.

SaFouNaldo reviewed (approved):
- Added Drawable Functions: Two new drawable functions have been added. Consider moving these to an image folder for better organization and maintainability.
- Help Profile Screen String: The help_profile_screen text is now managed in strings.xml, ensuring no hard-coded strings and better localization support.
Removed Unnecessary Imports: Cleaned up redundant imports to streamline the code and improve readability.
- Refactored ProfileScreen and MyStatsScreen: Both screens are now more concise, with fewer imports, enhancing overall efficiency.
- Enhanced Utils: Additional reusable functions have been added to Utils, reducing code duplication. These functions are thoroughly tested in UtilsTest.
- Updated MainActivity Parameters: Adjusted parameters in MainActivity to align with recent changes, promoting better functionality and consistency.
Everything looks good to go! Approved for merge.

ThomasMPont commented on gradle/libs.versions.toml:
```diff
@@ -7,6 +7,7 @@ json = "20240303"
 kotlin = "1.9.0"
 gms = "4.4.2"
 ktfmt = "0.17.0"
+sonar = "4.4.1.3373"
```
This was a good initiative, but you should have done the fix in a separate PR. This commit does not belong here.

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/ProfileScreen.kt:
```diff
@@ -63,199 +31,58 @@ fun ProfileScreen(
-              else colorResource(R.color.dark_gray).copy(alpha = .5f),
-          modifier = Modifier.padding(horizontal = 8.dp).testTag(letter.toString()))
-    }
+  ScaffoldMainScreen(
```
It's crazy how much simpler this reads now. Wow!

ThomasMPont commented on app/src/androidTest/java/com/github/se/signify/ui/screens/UtilsTest.kt:
```diff
@@ -222,6 +240,193 @@ class UtilsTest {
     assert(clicked)
   }
 
+  @Test
```
No comment, these tests are great.

ThomasMPont commented on app/src/androidTest/java/com/github/se/signify/ui/screens/profile/MyStatsScreenTest.kt:
```diff
@@ -40,46 +46,58 @@ class MyStatsScreenTest {
-    composeTestRule.onNodeWithTag("flameIcon").assertIsDisplayed()
-    composeTestRule.onNodeWithTag("numberOfDays").assertIsDisplayed()
-    composeTestRule.onNodeWithTag("numberOfDays").assertTextEquals("10 days")
+    // Verify top information are displayed correctly
```
It looks like you've added checks here. Don't get too distracted :) , sometimes you just have to let it be.

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/MyStatsScreen.kt:
What an effective refactor!

ThomasMPont reviewed (changes requested):
## Additions

Collecting elements from all our screens to define standard common elements was a gargantuan task. Good job getting to the end of it!

You have given future developers just enough freedom to be able to build screens without straying too far off the app's look.

## Requested Changes

I have requested changes by commenting on files. These comments are labled with the same title as this section. Please have a look at them.

## Description

Your description is both clear and complete.
Use `## ...` rather than `**...**` to define markdown subsections.
If you link to related tasks, make sure to include all of them. You probably just forgot to update that part of the descrition as new tasks were added to the PR.

## Tests

Tests look clean and achieve their purpose.

## Commits

Good job splitting this consequential PR into many commits. This removes some of the strain of dealing with big additions like this one.

The commits,
- > run of the gradlew command
- > run ktfmtFormat
- > run gradle function

don't match your use of conventional commits. Additionally, commits should reflect what was added more importantly than what you used.

The commit,
> refactor(ProfileScreen): use common function from Utils

add resource files for some reason. I feel like these could have been added in a separate commit, or at least been mentioned somewhere.

Try to resist the urge to fix blatant mistakes, such as typos or missing lines, in your current commit. You should always commit files separately, stash your changes or keep changes for a future task, rather that fixing random issues in unrelated commits.
I've pointed out two occurences of this in file comments.

## Future additions

Now that these common elements are defined. We must refactor all existing screens to use them. Tasks achieving this have been created in the sprint backlog.

## Conclusion

Despite the long review, this PR is exemplary, and deserves a merge as soon as its requested changes are addressed.

NBCCF025 commented on app/src/main/java/com/github/se/signify/ui/Utils.kt:
```diff
@@ -46,17 +75,24 @@ import com.github.se.signify.R
+    }
 }
 
 /**
```
To have something similar to the `UtilIconButton()`, I renamed it `UtilTextButton`.

NBCCF025 commented on app/src/main/java/com/github/se/signify/ui/Utils.kt:
```diff
@@ -151,64 +211,577 @@ fun UtilButton(
+fun StatisticsRow(
+    rowTestTag: String,
+    lineText: String,
+    lineTextTag: String,
```
This is not a typo, it it the testTag of the corresponding lineText, but I wouldn't name it lineTextTestTag and just choose to add the word `Tag` to keep small. But if you want I can rename it by lineTestTag.

ThomasMPont replied:
Being verbose would probably be better here, but I don't think it's worth changing at this point.

NBCCF025 commented:
I have made some changes that were requested, now it should be good.

ThomasMPont commented:
> I have made some changes that were requested, now it should be good.

You did. Nice work. I'll go ahead and merge everything.
